File: executioner/languages/docker.ts
import fs from 'fs';
import path from 'path';
import cp from 'child_process';
import { v4 } from 'uuid';

import { readFile, writeFile } from 'fs/promises';

type LanguageOptions = {
  language: string;
  entrypointFile: string;
  timeLimitSeconds?: number;
  memoryLimitMb?: number;
  interactive?: string;
  interactorSource?: string;
  checkerSource?: string;
  [key: string]: string | number | undefined;
};

type File = {
  id?: string;
  filename: string;
  content: string;
  metadata?: any;
};

type ExecutionResult = {
  stdout: string;
  stderr: string;
  time: number | null;
  memoryKb: number | null;
  verdict: string | null;
  metadata: any;
};

// Extra MB added to --memory so bash/time/timeout overhead doesn't eat into
// the problem's stated limit. The soft RSS check below enforces the actual limit.
const MEMORY_OVERHEAD_MB = 16;

const CPU_LIMIT_PER_EXECUTION = process.env.CPU_LIMIT_PER_EXECUTION;
const DEFAULT_MEMORY_LIMIT = process.env.MEMORY_LIMIT_PER_EXECUTION ?? '256m';
const DEFAULT_TIME_LIMIT_SECONDS = parseFloat(process.env.EXECUTION_TIMEOUT ?? '10');
const ENABLE_NETWORK_IN_EXECUTION = process.env.ENABLE_NETWORK_IN_EXECUTION === 'true';
const IMAGE_BASE = process.env.IMAGE_BASE || 'ghcr.io/42dotmk/colosseum-executioner-';
const WORKDIR = process.env.WORKDIR || '_work';

const normalizeRuntimeLanguage = (language: string) => {
  if (!language) {
    return language;
  }

  const normalized = language.toLowerCase();

  if (normalized === 'cpp' || normalized === 'c++' || normalized === 'cxx') {
    return 'gcc';
  }

  return normalized;
};

if (!fs.existsSync(WORKDIR)) {
  fs.mkdirSync(WORKDIR);
}


const readIfExists = async (path: string) => {
  if (fs.existsSync(path)) {
    return (await readFile(path)).toString();
  }
  return '';
}

export const execute = async (files: File[], input: File[], options: LanguageOptions) => {
  const id = v4();

  const subWorkspace = path.join(WORKDIR, id);
  const srcDir = path.resolve(path.join(subWorkspace, "src"));
  const inputDir = path.resolve(path.join(subWorkspace, "input"));
  const outputDir = path.resolve(path.join(subWorkspace, "output"));
  const lang = normalizeRuntimeLanguage(options.language);

  if (!fs.existsSync(subWorkspace)) {
    fs.mkdirSync(subWorkspace);
  }

  if (!fs.existsSync(srcDir)) {
    fs.mkdirSync(srcDir);
  }

  if (!fs.existsSync(inputDir)) {
    fs.mkdirSync(inputDir);
  }

  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir);
  }

  const timeFilename = `time`;
  const stdoutFilename = `stdout`;
  const compileStdoutFilename = `compile.stdout`;
  const compileStderrFilename = `compile.stderr`;
  const stderrFilename = `stderr`;

  const timePath = path.resolve(path.join(subWorkspace, timeFilename));
  const stdoutPath = path.resolve(path.join(subWorkspace, stdoutFilename));
  const stderrPath = path.resolve(path.join(subWorkspace, stderrFilename));
  const compileStdoutPath = path.resolve(path.join(subWorkspace, compileStdoutFilename));
  const compileStderrPath = path.resolve(path.join(subWorkspace, compileStderrFilename));

  for (const file of files) {
    const filePath = path.resolve(path.join(srcDir, file.filename));
    console.log(`Writing file to ${filePath}`)
    await writeFile(filePath, file.content);
  }

  // Write interactor and checker sources for interactive problems
  if (options.interactive === '1') {
    if (options.interactorSource) {
      await writeFile(path.resolve(path.join(srcDir, 'interactor.cpp')), options.interactorSource);
    }
    if (options.checkerSource) {
      await writeFile(path.resolve(path.join(srcDir, 'checker.cpp')), options.checkerSource);
    }
  }

  for (const file of input) {
    const filePath = path.resolve(path.join(inputDir, file.filename));
    await writeFile(filePath, file.content);
  }

  try {
    return await new Promise<ExecutionResult[]>(async (resolve) => {
      await writeFile(timePath, "");
      await writeFile(stdoutPath, "");
      await writeFile(stderrPath, "");
      await writeFile(compileStdoutPath, "");
      await writeFile(compileStderrPath, "");

      const extraArgs = [];

      if (CPU_LIMIT_PER_EXECUTION) {
        extraArgs.push(`--cpus=${CPU_LIMIT_PER_EXECUTION}`);
      }

      const effectiveMemoryMb = options.memoryLimitMb
        ? options.memoryLimitMb + MEMORY_OVERHEAD_MB
        : DEFAULT_MEMORY_LIMIT;
      const effectiveMemoryStr = typeof effectiveMemoryMb === 'number'
        ? `${effectiveMemoryMb}m`
        : effectiveMemoryMb;
      extraArgs.push(`--memory=${effectiveMemoryStr}`);
      // Setting memory-swap equal to memory disables swap, so OOM kill is
      // immediate and deterministic rather than relying on the host's swap space.
      extraArgs.push(`--memory-swap=${effectiveMemoryStr}`);

      const effectiveTimeout = options.timeLimitSeconds ?? DEFAULT_TIME_LIMIT_SECONDS;
      extraArgs.push(`-e`, `EXECUTION_TIMEOUT=${effectiveTimeout}`);

      if (!ENABLE_NETWORK_IN_EXECUTION) {
        extraArgs.push("--network=none");
      }

      if (options.interactive === '1') {
        extraArgs.push(`-e`, `INTERACTIVE=1`);
        extraArgs.push(`-e`, `SOLUTION_FILE=${options.entrypointFile}`);
      }

      const args = [
        'run',
        "--rm",
        "-v",
        `${srcDir}:/exc/src`,
        "-v",
        `${inputDir}:/exc/input`,
        "-v",
        `${outputDir}:/exc/output`,
        "-v",
        `${timePath}:/exc/${timeFilename}`,
        "-v",
        `${stdoutPath}:/exc/${stdoutFilename}`,
        "-v",
        `${stderrPath}:/exc/${stderrFilename}`,
        "-v",
        `${compileStdoutPath}:/exc/${compileStdoutFilename}`,
        "-v",
        `${compileStderrPath}:/exc/${compileStderrFilename}`,
        ...extraArgs,
        "-i",
        `${IMAGE_BASE}${lang}`
      ];

      const child = cp.spawn('docker', args);
      console.log(child.spawnargs.join(' '))

      let runtimeStdout = '';
      let runtimeStderr = '';
      let spawnError: string | null = null;

      child.stdout.on('data', (data) => {
        console.log(`stdout: ${data}`);
        runtimeStdout += data;
      });

      child.stderr.on('data', (data) => {
        console.error(`stderr: ${data}`);
        runtimeStderr += data;
      });

      child.on('error', (err) => {
        spawnError = err instanceof Error ? err.message : String(err);
        console.error(`docker spawn error: ${spawnError}`);
      });

      child.on('close', async (code) => {
        console.log(`child process exited with code ${code}`);

        const output = [];
        const dockerErrorDetails = [spawnError, runtimeStderr.trim(), runtimeStdout.trim()]
          .filter(Boolean)
          .join("\n")
          .trim();

        for (const inp of input) {
          const stdoutPath = path.resolve(path.join(outputDir, `${inp.filename}.stdout`));
          const stderrPath = path.resolve(path.join(outputDir, `${inp.filename}.stderr`));
          const timePath = path.resolve(path.join(outputDir, `${inp.filename}.time`));
          const verdictPath = path.resolve(path.join(outputDir, `${inp.filename}.verdict`));
          const stdout = await readIfExists(stdoutPath);

          let stderr = await readIfExists(stderrPath);

          if (!stdout && !stderr && code !== 0) {
            stderr = dockerErrorDetails
              ? `Execution runtime failed (exit code ${code}):\n${dockerErrorDetails}`
              : `Execution runtime failed (exit code ${code})`;
          }

          const timeContent = await readIfExists(timePath);
          let parsedTime: number | null = null;
          let parsedMemoryKb: number | null = null;
          if (timeContent) {
            // GNU time may prepend "Command exited with non-zero status N\n"
            // on non-zero exits. Scan lines in reverse for the "seconds kb" line.
            const lines = timeContent.trim().split('\n').reverse();
            for (const line of lines) {
              const parts = line.trim().split(/\s+/);
              if (parts.length === 2) {
                const t = parseFloat(parts[0]);
                const m = parseInt(parts[1], 10);
                if (!isNaN(t) && !isNaN(m)) {
                  parsedTime = t;
                  parsedMemoryKb = m;
                  break;
                }
              }
            }
            if (parsedTime === null) {
              console.error(`Failed to parse time from: ${timeContent}`);
            }
          }

          const verdictContent = await readIfExists(verdictPath);
          let verdict = verdictContent.trim() || (code === 137 ? 'MLE' : null);

          const memoryLimitKb = options.memoryLimitMb != null ? options.memoryLimitMb * 1024 : null;

          // Soft RSS check: program completed cleanly but exceeded the problem's
          // memory limit. The overhead buffer means Docker won't hard-kill it in
          // that case, so we enforce the limit here.
          let isSoftMle = false;
          if (verdict === 'OK' && memoryLimitKb !== null && parsedMemoryKb !== null && parsedMemoryKb > memoryLimitKb) {
            verdict = 'MLE';
            isSoftMle = true;
          }

          // Hard MLE (OOM kill): the RSS at kill time reflects the Docker ceiling
          // (memoryLimitMb + MEMORY_OVERHEAD_MB), not the program's real peak, so
          // it's misleadingly high. Cap at the problem limit so callers see a clean
          // lower-bound value ("used at least N KB") rather than Docker internals.
          if (verdict === 'MLE' && !isSoftMle && memoryLimitKb !== null && parsedMemoryKb !== null && parsedMemoryKb > memoryLimitKb) {
            parsedMemoryKb = memoryLimitKb;
          }

          output.push({
            stdout: stdout.toString(),
            stderr: stderr.toString(),
            time: parsedTime,
            memoryKb: parsedMemoryKb,
            verdict,
            metadata: inp.metadata,
          });
        }

        resolve(output);
      });

      console.log('child', child.pid);
    });
  } catch (e) {
    console.error(e);
  } finally {
    if (fs.existsSync(subWorkspace)) {
      fs.rmSync(subWorkspace, { recursive: true });
    }
  }
};

File: executioner/languages/docker.go
package languages

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/google/uuid"
	"github.com/joho/godotenv"
)

type LanguageOptions struct {
	Language       string            `json:"language"`
	EntrypointFile string            `json:"entrypointFile"`
	Options        map[string]string `json:"options"`
}

type File struct {
	ID       string                `json:"id,omitempty"`
	Filename string                 `json:"filename"`
	Content  string                 `json:"content"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}
type ExecutionResult struct{
	ID string `json:"id"`
	Stdout string `json:"stdout"`
	Stderr string `json:"stderr"`
	Time int `json:"time"`
	Metadata map[string]interface{} `json:"metadata"`
}
var (
	CPU_LIMIT_PER_EXECUTION string
	MEMORY_LIMIT_PER_EXECUTION string
	ENALBE_NETWORK_IN_EXECUTION string
	IMAGE_BASE string
	WORKDIR string
)

const (
	timeFilename string = "time"
	stdoutFilename string = "stdout"
	StderrFilename string = "stderr"
	compileStdoutFilename string = "compile.stdout"
	compileStderrFilename string = "compile.stderr"

)

func Init (){
	err := godotenv.Load()
	if err != nil {
		fmt.Printf("Failed to load env: %s\n", err)
	}
	CPU_LIMIT_PER_EXECUTION = os.Getenv("CPU_LIMIT_PER_EXECUTION")
	MEMORY_LIMIT_PER_EXECUTION = os.Getenv("MEMORY_LIMIT_PER_EXECUTION")
	ENALBE_NETWORK_IN_EXECUTION = os.Getenv("ENALBE_NETWORK_IN_EXECUTION")
	IMAGE_BASE = os.Getenv("IMAGE_BASE")
	WORKDIR = os.Getenv("WORKDIR")
	if WORKDIR == "" {
		WORKDIR = "_work"
	}
	
}

func parseDuration(duration string) (int, error) {
	if duration == "" {
		return 0, fmt.Errorf("Received invalid duration '%s'", duration)
	}
	re := regexp.MustCompile(`(\d+)m(\d+(?:\.\d+)?)s`)
	match := re.FindStringSubmatch(duration)
	if match == nil {
		return 0, fmt.Errorf("Invalid duration format")
	}

	minutes, err := strconv.Atoi(match[1])
	if err != nil {
		return 0, err
	}
	seconds, err := strconv.ParseFloat(match[2], 64)
	if err != nil {
		return 0, err
	}

	return minutes*60 + int(seconds), nil
}

func readIfExist(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return string(content), nil
}

func StartCodeContainer(sources []File, input []File, options LanguageOptions) (string, error) {
	id := uuid.New()

	files:= []File{}
			
	subWorkspace := filepath.Join(WORKDIR, id.String())
	srcDir := filepath.Join(subWorkspace, "src") 
	inputDir := filepath.Join(subWorkspace, "input")
	outputDir := filepath.Join(subWorkspace, "output")
	lang := options.Language


	_, err := os.ReadDir(subWorkspace)
	if err != nil {
		err := os.MkdirAll(subWorkspace, 0777)
		if err != nil {
			return "", err
		}
	}
	_, err = os.ReadDir(srcDir)
	if err != nil {
		err := os.MkdirAll(srcDir, 0777)
		if err != nil {
			return "", err
		}
	}
	_, err = os.ReadDir(inputDir)
	if err != nil {
		err := os.MkdirAll(inputDir, 0777)
		if err != nil {
			return "", err
		}
	}
	_, err = os.ReadDir(outputDir)
	if err != nil {
		err := os.MkdirAll(outputDir, 0777)
		if err != nil {
			return "", err
		}
	}

	for _, source := range sources {
		
		content := source.Content
		err := writeFileToDir(srcDir, source)
		if err != nil {
			return "", fmt.Errorf("Failed to write file: %s", err)
		}
		files = append(files, File{
			ID:       id.String(),
			Filename: source.Filename,
			Content: string(content[:]),
		})
	}
	for _ , inp := range input {
		content := inp.Content
		err := writeFileToDir(inputDir, inp)
		if err != nil {
			return "", fmt.Errorf("Failed to write file: %s", err)
		}
		files = append(files, File{
			ID:       id.String(),
			Filename: inp.Filename,
			Content: string(content[:]),
			Metadata: inp.Metadata,
		})
	}

	timePath := filepath.Join(subWorkspace,timeFilename)
	stdoutPath := filepath.Join(subWorkspace, stdoutFilename)
	stderrPath := filepath.Join(subWorkspace, StderrFilename)
	compileStdoutPath := filepath.Join(subWorkspace, compileStdoutFilename)
	compileStderrPath := filepath.Join(subWorkspace, compileStderrFilename)

	filesToCreate := []string{timePath, stdoutPath, stderrPath, compileStdoutPath, compileStderrPath}
	for _, filePath := range filesToCreate {
		_, err := os.Stat(filePath)
		if os.IsNotExist(err) {
			err = os.WriteFile(filePath, []byte(""), 0644)
			if err != nil {
				return "", fmt.Errorf("Failed to create file %s: %s", filePath, err)
			}
		}
	}

	var extraArgs []string

	if CPU_LIMIT_PER_EXECUTION != "" {
		extraArgs = append(extraArgs, "--cpus", CPU_LIMIT_PER_EXECUTION)
	}
	if MEMORY_LIMIT_PER_EXECUTION != "" {
		extraArgs = append(extraArgs, "--memory", MEMORY_LIMIT_PER_EXECUTION)
		
	}
	if ENALBE_NETWORK_IN_EXECUTION == "" {
		extraArgs = append(extraArgs, "--network", "none")
	}

	args := []string{
		"run",
		"--rm",
		"-v", fmt.Sprintf("%s:/exc/src", srcDir),
		"-v", fmt.Sprintf("%s:/exc/input", inputDir),
		"-v", fmt.Sprintf("%s:/exc/output", outputDir),
		"-v", fmt.Sprintf("%s:/exc/%s", timePath, timeFilename),
		"-v", fmt.Sprintf("%s:/exc/%s", stdoutPath, stdoutFilename),
		"-v", fmt.Sprintf("%s:/exc/%s", stderrPath, StderrFilename),
		"-v", fmt.Sprintf("%s:/exc/%s", compileStdoutPath, compileStdoutFilename),
		"-v", fmt.Sprintf("%s:/exc/%s", compileStderrPath, compileStderrFilename),
	}
	args = append(args, extraArgs...)
	args = append(args, "-i", IMAGE_BASE+lang, options.EntrypointFile)

	cmdDocker := exec.Command("docker", args...)

	fmt.Printf("Command: %s\n", cmdDocker.String())

	stdout,err:= cmdDocker.StdoutPipe()
	if err != nil {
		return "", fmt.Errorf("Failed to get stdout pipe: %s", err)
	}
	fmt.Println(stdout)
	stderr,err:= cmdDocker.StderrPipe()
	if err != nil {
		return "", fmt.Errorf("Failed to get stderr pipe: %s", err)
	}
	fmt.Println(stderr)
	err = cmdDocker.Start()

	if err != nil {
		return "", fmt.Errorf("Failed to start docker: %s", err)
	}
	
	// read stdout and stderr into string

	stdoutput, err := io.ReadAll(stdout)
	if err != nil {
		return "", fmt.Errorf("Failed to read stdout: %s", err)
	}

	errOutput, err := io.ReadAll(stderr)
	if err != nil {
		return "", fmt.Errorf("Failed to read stderr: %s", err)
	}

	if len(errOutput) > 0 {
		return "", fmt.Errorf("Failed to execute: %s", string(errOutput))
	}


	stdFiles := []File{}
	suffixes := []string{".stdout", ".stderr", ".time"}
	for _, inp := range input {
		for _, suffix := range suffixes {
			filePath := filepath.Join(outputDir, inp.Filename+suffix)
			content, err := os.ReadFile(filePath)
			if err != nil {
				return "", fmt.Errorf("Failed to read file %s: %s", filePath, err)
			}
			stdFiles = append(stdFiles, File{
				ID:       id.String(),
				Filename: filepath.Base(filePath),
				Content:  string(content),
				Metadata: inp.Metadata,
			})
		}
	}


	fmt.Printf("Stdout: %s\n", string(stdoutput))
	fmt.Printf("Stderr: %s\n", string(errOutput))


	err = cmdDocker.Wait()
	if err != nil {
		return "", fmt.Errorf("Failed to wait for docker: %s", err)
	}
		
	output,err := processOutput(stdFiles)
	if err != nil{
		return "", fmt.Errorf("Failed to process output: %s", err)
	}
	return output, nil
}


func processOutput(stdFiles []File) (string, error) {

	grouped := make(map[string]map[string]File)
	for _, file := range stdFiles {
		ext := filepath.Ext(file.Filename)
		baseName := strings.TrimSuffix(file.Filename, ext)
		if grouped[baseName] == nil {
			grouped[baseName] = make(map[string]File)
		}
		grouped[baseName][ext] = file
	}

	var results []ExecutionResult
	for base, filesMap := range grouped {
		stdout := ""
		stderr := ""
		timeContent := ""

		if file, ok := filesMap[".stdout"]; ok {
			stdout = file.Content
		} else {
			return "", fmt.Errorf("Missing stdout for %s", base)
		}
		if file, ok := filesMap[".stderr"]; ok {
			stderr = file.Content
		} else {
			return "", fmt.Errorf("Missing stderr for %s", base)
		}
		if file, ok := filesMap[".time"]; ok {
			timeContent = file.Content
		} else {
			return "", fmt.Errorf("Missing time for %s", base)
		}

		parsedTime := 0
		if timeContent != "" {
			var timeSplits [][]string
			for _, t := range regexp.MustCompile(`\r?\n`).Split(timeContent, -1) {
				trimmed := strings.TrimSpace(t)
				if trimmed != "" {
					timeSplits = append(timeSplits, strings.Split(trimmed, "\t"))
				}
			}
			if len(timeSplits) > 0 && len(timeSplits[0]) > 1 {
				var err error
				parsedTime, err = parseDuration(timeSplits[0][1])
				if err != nil {
					return "", fmt.Errorf("Failed to parse duration for %s: %s", base, err)
				}
			}
		}

		var fileID string
		var metadata map[string]interface{}
		if file, ok := filesMap[".stdout"]; ok {
			fileID = file.ID
			metadata = file.Metadata
		}

		result := ExecutionResult{
			ID:       fileID,
			Stdout:   strings.TrimSpace(stdout),
			Stderr:   strings.TrimSpace(stderr),
			Time:     parsedTime,
			Metadata: metadata,
		}
		results = append(results, result)
	}

	jsonOutput, err := json.Marshal(results)
	if err != nil {
		return "", fmt.Errorf("Failed to marshal output: %s", err)
	}
	return string(jsonOutput), nil
}


func writeFileToDir(path string, input File) error {
	err := os.WriteFile(filepath.Join(path, input.Filename), []byte(input.Content), 0777)
	if err != nil {
		return fmt.Errorf("Failed to write file: %s", err)
	}
	return nil
}